### PR TITLE
add fontawesome mapper

### PR DESCRIPTION
**Pull request by printminion-co**
Preconditions:
* you need to have a FontAwesome Pro token in order to use the icons in your project https://fontawesome.com/account#pro-package-tokens

# Installation
`FONTAWESOME_PACKAGE_TOKEN=<TOKEN> npm install`

### CODE REVIEW

File: iconsMapper.ts [hunk 1]
```diff
@@ -8,15 +23,66 @@ const mdiViewBox = {
...
     icons: { ...mdiIcons },
     viewBoxes: {},
+    stats: {},
   };
 
   for (const icon in result.icons) {
     // add the viewBox for the icon in order to build the template correctly
     result.viewBoxes[icon] = mdiViewBox;
+    result.stats[icon] = MappingStatus.ORIGINAL;
+    if (!mapping.icons.hasOwnProperty(icon)) {
```
Comment (thlehmann-ionos) [inline iconsMapper.ts]: 1. This check will bail out for unmapped icons too, making the next check that logs and collects "stats" redundant/no-op
    An unmapped icon would not be logged and not show up in the stats.
    I assume the purpose was to skip Javascript internal properties. They should not be present with `for..in` anyway.
    I think it is safe to remove this condition.
2. `hasOwnProperty()` is [deprecated][mdn-hasownprop], please use `Object.hasOwn()` instead

Altenative approach: make two diff (both ways) and one intersection of the keys of `result.icons` and `mapping.icons`, dump the differences (unmapped or unknown) and process the intersection.

[mdn-hasownprop]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn

File: iconsMapper.ts [hunk 2]
```diff
@@ -8,15 +23,66 @@ const mdiViewBox = {
 
 const remap = reMapIcons();
 
+/**
+ * Get the icon lookup from the icon string
```
Comment (thlehmann-ionos) [inline iconsMapper.ts]: This is weirdly phrased, please rephrase.
Please explain at the parameter doc what's the expected input format

File: iconsMapper.ts [hunk 3]
```diff
@@ -8,15 +23,66 @@ const mdiViewBox = {
...
+ * @param icon
+ * @returns IconLookup
+ */
+function getIconName(icon: string): IconLookup {
```
Comment (thlehmann-ionos) [inline iconsMapper.ts]: `IconLookup` is not defined

**Comment by thlehmann-ionos** [comment]
Please update/change/rewrite the readme a bit. Keep it _short_ .

Relevant info: how to build, how to install, how to publish, what's different (i.e. it's no longer material design icons).

File: contactsheet.ts [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+      mdiIconData.svgPathData,
+      viewBoxMdi,
+    );
+
+    const viewBoxDefault = {
+      width: 24,
+      height: 24,
+    };
+
+    const viewBox = viewBoxes[mdiIconId] || viewBoxDefault;
```
Comment (thlehmann-ionos) [inline contactsheet.ts]: Why's there a default, all icons should already have a value (the same value) in the mapping, no?

**Review by thlehmann-ionos — CHANGES_REQUESTED** [review]
Intermediate review feedback. More testing to be done.

File: contactsheet.ts (hunk 4 above)
Comment (printminion-co) [reply contactsheet.ts]: yes... ..initially I hadn't set mdi icons viewBox... ...this line can be changed

File: iconsMapper.ts (hunk 3 above)
Comment (printminion-co) [reply iconsMapper.ts]: is part of fontawesome package (don't remeber whitch).. ..your IDE should resolve it

File: iconsMapper.ts (hunk 1 above)
Comment (printminion-co) [reply iconsMapper.ts]: _"An unmapped icon would not be logged and not show up in the stats."_ - all stats are userful for building the contactsheet. To help the map person to find faster the map errors

File: iconsMapper.ts (hunk 3 above)
Comment (thlehmann-ionos) [reply iconsMapper.ts]: It has to be imported:
image

```
$ node_modules/.bin/tsc
...
iconsMapper.ts:35:37 - error TS2304: Cannot find name 'IconLookup'.

35 function getIconName(icon: string): IconLookup {
                                       ~~~~~~~~~~
```

Comment (printminion-co) [reply iconsMapper.ts]: 🙈 my bad... ..I had it imported.. ..probably killed due rebasing

File: iconsMapper.ts (hunk 1 above)
Comment (thlehmann-ionos) [reply iconsMapper.ts]: I understand what the stats are for.

However, the `hasOwnProperty()` and the following `!newIconName` are redundant because they both test for undefined field. 

I now see that the 2nd branch also tests for an empty string, in that case it's not redundant and the state is sensible. 

I'd propose to change the log message and maybe change to `newIconName === ''` to make this clearer or to test for a name pattern via a RegExp if this is about preventing malformed/unexpected icon names. Maybe we can even lookup the name in the library?

File: mapping.ts [hunk 5]
```diff
@@ -6,6 +6,88 @@
...
+    mdiAlertOutline: 'fas-exclamation-triangle',
+    mdiArrowDown: 'fas-arrow-down',
+    mdiArrowLeft: 'fas-arrow-left',
+    mdiArrowRight: 'fas-arrow-right',
+    mdiArrowUp: 'fas-arrow-up',
+    mdiBackupRestore: 'fas-undo',
+    mdiCalendarRange: 'fas-calendar-alt',
+    mdiChartPie: 'fas-pie-chart',
+    mdiCheck: 'fas-check',
+    mdiCheckBold: 'here is no equivalent in fontawesome',
```
Comment (thlehmann-ionos) [inline mapping.ts]: It's the tick icon that appears after successfully sharing a file. In our screen designs it looks like the "regular" tick icon. Let's map to that instead.

**Comment by thlehmann-ionos** [comment]
See review feedback in branch `tl/review/mk/poc/remap-to-fontawesome` too. Consider changing the mapper script to Javascript because there are many Typescript errors, which would take more effort to fix than to treat this as Javascript. Feel free to property type though.

File: mapping.ts [hunk 6]
```diff
@@ -7,87 +7,87 @@
...
     mdiFilter: 'fas-filter',
     mdiFolder: 'fas-folder',
     mdiFolderOpen: 'fas-folder-open',
-    mdiFormatListBulletedSquare: 'fas-list-ul',
-    mdiFullscreen: 'fas-maximize',
-    mdiFullscreenExit: 'fas-minimize',
+    mdiFormatListBulletedSquare: 'fasr-list',
+    mdiFullscreen: 'fass-expand',
+    mdiFullscreenExit: 'fass-compress',
     mdiImageEdit: null,
```
Comment (thlehmann-ionos) [inline mapping.ts:55]: This one is still `null`.

Comment (printminion-co) [reply mapping.ts:55]: it's a demo to show a fallback to MDI icons in case there is no similar asset in FA icons

Comment (printminion-co) [reply mapping.ts:55]: run ` npm run contactsheet` to see the output in `docs/mapping.html`
image

Comment (thlehmann-ionos) [reply mapping.ts:55]: I saw it, but it was not obvious that this was a demo since it was not added by a commit marked with `[DEMO]`.

**Comment by thlehmann-ionos** [comment]
I had a 2nd quick look at it, looked good to me.
